package/gluon-radv-filterd/src/gluon-radv-filterd.c

```diff
@@ -0,0 +1,577 @@
+	foreach(router, G.routers) {
+		if (router->tq == 0)
+			break;
+	}
+	if (router != NULL) {
+		// rate local routers (if present) the highest
+		snprintf(path, PATH_MAX, TRANSTABLE_LOCAL, G.mesh_iface);
+		f = fopen(path, "r");
+		// skip header
+		while (fgetc(f) != '\n');
+		while (fgetc(f) != '\n');
+		while (fscanf(f, " * " F_MAC " [%*5s] %*f", F_MAC_VAR_REF(mac_a)) == 6) {
+			foreach(router, G.routers) {
+				if (!memcmp(router->src, mac_a, sizeof(macaddr_t))) {
+					DEBUG_MSG("Found router " F_MAC " in transtable_local, assigning TQ %d", F_MAC_VAR(router->src), LOCAL_TQ);
+					router->tq = LOCAL_TQ;
+					G.max_tq = LOCAL_TQ;
+					break; // foreach
+				}
+			}
+		}
+		if (!feof(f)) {
+			getline(&line, &len, f);
+			fprintf(stderr, "Parsing transtable_local aborted at this line: %s\n", line);
+		}
+		fclose(f);
+	}
+
+	foreach(router, G.routers) {
+		if (router->tq == 0) {
+			if (!memcmp(router->originator, unspec, sizeof(unspec)))
+				fprintf(stderr,
+					"Unable to find router " F_MAC " in transtable_{global,local}\n",
+					F_MAC_VAR(router->src));
+			else
+				fprintf(stderr,
+					"Unable to find TQ for originator " F_MAC " (router " F_MAC ")\n",
+					F_MAC_VAR(router->originator),
+					F_MAC_VAR(router->src));
+		}
+	}
+
+	free(line);
+}
+
+static int fork_execvp_timeout(struct timespec *timeout, const char *file, const char *const argv[]) {
+	int ret;
+	pid_t child;
+	siginfo_t info;
+	sigset_t signals, oldsignals;
+	sigemptyset(&signals);
+	sigaddset(&signals, SIGCHLD);
+
+	sigprocmask(SIG_BLOCK, &signals, &oldsignals);
+	child = fork();
+	if (child == 0) {
+		sigprocmask(SIG_SETMASK, &oldsignals, NULL);
+		// casting discards const, but should be safe
+		// (see http://stackoverflow.com/q/36925388)
+		execvp(file, (char**) argv);
+		fprintf(stderr, "can't execvp(\"%s\", ...): %s\n", file, strerror(errno));
+		_exit(1);
+	}
+	else if (child < 0) {
+		perror("Failed to fork()");
+		return -1;
+	}
+
+	ret = sigtimedwait(&signals, &info, timeout);
+	sigprocmask(SIG_SETMASK, &oldsignals, NULL);
+
+	if (ret == SIGCHLD) {
+		if (info.si_pid != child) {
+			cleanup();
+			error_message(1, 0,
+				"BUG: We received a SIGCHLD from a child we didn't spawn (expected PID %d, got %d)",
+				child, info.si_pid);
+		}
+
+		waitpid(child, NULL, 0);
+
+		return info.si_status;
+	}
+
+	if (ret < 0 && errno == EAGAIN)
+		error_message(0, 0, "warning: child %d took too long, killing", child);
+	else if (ret < 0)
+		warn_errno("sigtimedwait failed, killing child");
+	else
+		error_message(1, 0,
+				"BUG: sigtimedwait() returned some other signal than SIGCHLD: %d",
+				ret);
+
+	kill(child, SIGKILL);
+	kill(child, SIGCONT);
+	waitpid(child, NULL, 0);
+	return -1;
+}
+
+static void update_ebtables() {
+	struct timespec timeout = {
+		.tv_nsec = EBTABLES_TIMEOUT,
+	};
+	char mac[F_MAC_LEN + 1];
+	struct router *router;
+
+	if (G.best_router && G.best_router->tq >= G.max_tq - G.hysteresis_thresh) {
```

Comment:
Here is a potential underflow. It should work for you because max_tq and tq are now uint16_t and  G.hysteresis_thresh must be smaller than 512 - but it is slightly odd and it makes problems in @belzebub40k's implementation (using uint8_t and unlimited G.hysteresis_thresh).

Maybe there could be a small static function which checks G.best_router and the hysteresis threshold: 

is_eligible:

* if !G.best_router
  - return True
* if G.max_tq < G.best_router->tq (yes, this should actually not happen)
  - return false
* if (G.max_tq - G.best_router->tq) < G.hysteresis_thresh
  - return false
* otherwise
  - return true